src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/Connection.php

```diff
@@ -0,0 +1,188 @@
+    {
+        $this->configuration = array_replace_recursive(self::DEFAULT_OPTIONS, $configuration);
+        $this->client = $client ?? new PubSubClient([]);
+    }
+
+    /**
+     * Creates a connection based on the DSN and options.
+     *
+     * Available options:
+     *
+     * * apiEndpoint: The hostname with optional port to use in place of the default service endpoint
+     * * projectId: The project ID from the Google Developer's Console
+     * * keyFile: The contents of the service account credentials.json file retrieved from the Google Developer's Console.
+     * * keyFilePath: The full path to your service account credentials .json file retrieved from the Google Developers Console.
+     * * requestTimeout: Seconds to wait before timing out the request. (Defaults: `0` with REST and `60` with gRPC)
+     * * retries: Number of retries for a failed request. (Default: `3`)
+     * * scopes: Scopes to be used for the request
+     * * quotaProject: Specifies a user project to bill for access charges associated with the request
+     * * topic: The options for the topic to publish messages to
+     * * subscription: The options of the subscription to receive messages from
+     * * auto_setup: Whether the queue should be created automatically during send / get (Default: true)
+     */
+    public static function fromDsn(string $dsn, array $options = []): self
+    {
+        if (false === $parsedUrl = parse_url($dsn)) {
+            throw new InvalidArgumentException(sprintf('The given Google Pub/Sub DSN "%s" is invalid.', $dsn));
+        }
+
+        $query = [];
+        if (isset($parsedUrl['query'])) {
+            parse_str($parsedUrl['query'], $query);
+        }
+
+        // Check for unsupported keys in options
+        $optionsExtraKeys = array_diff(array_keys($options), array_keys(self::DEFAULT_OPTIONS));
+        if (0 < \count($optionsExtraKeys)) {
+            throw new InvalidArgumentException(sprintf('Unknown option found: [%s]. Allowed options are [%s].', implode(', ', $optionsExtraKeys), implode(', ', array_keys(self::DEFAULT_OPTIONS))));
+        }
+
+        // Check for unsupported keys in query
+        $queryExtraKeys = array_diff(array_keys($query), array_keys(self::DEFAULT_OPTIONS));
+        if (0 < \count($queryExtraKeys)) {
+            throw new InvalidArgumentException(sprintf('Unknown option found in DSN: [%s]. Allowed options are [%s].', implode(', ', $queryExtraKeys), implode(', ', array_keys(self::DEFAULT_OPTIONS))));
+        }
+
+        $options       = $query + $options + self::DEFAULT_OPTIONS;
+        $configuration = [
+            'topic' => $options['topic'],
+            'subscription' => $options['subscription'],
+            'auto_setup' => (bool) $options['auto_setup'],
+        ];
+
+        $clientConfiguration = [
+            'projectId' => ltrim($parsedUrl['path'] ?? '', '/') ?? self::DEFAULT_OPTIONS['projectId'],
```

Comment:
This doesn't look right. The return value of `ltrim` is never `null`, so the second null coalescing operator won't do anything.